### DISCUSSION TITLE
boost:bind的函数类型不对

### DISCUSSION THREAD

**Discussion by icooperlee** [chenshuo/muduo#583] [General]
Channel.h

class Channel : boost::noncopyable
{
 public:
  typedef boost::function<void()> EventCallback;
  _**typedef boost::function<void(Timestamp)> ReadEventCallback;**_

  Channel(EventLoop* loop, int fd);
  ~Channel();

  void handleEvent(Timestamp receiveTime);
  **_void setReadCallback(const ReadEventCallback& cb)
  { readCallback_ = cb; }**
  void setWriteCallback(const EventCallback& cb)
  { writeCallback_ = cb; }
  void setCloseCallback(const EventCallback& cb)
  { closeCallback_ = cb; }
  void setErrorCallback(const EventCallback& cb)
  { errorCallback_ = cb; }

EventLoop.cc

EventLoop::EventLoop()
  : looping_(false),
    quit_(false),
    eventHandling_(false),
    callingPendingFunctors_(false),
    iteration_(0),
    threadId_(CurrentThread::tid()),
    poller_(Poller::newDefaultPoller(this)),
    timerQueue_(new TimerQueue(this)),
    wakeupFd_(createEventfd()),
    wakeupChannel_(new Channel(this, wakeupFd_)),
    currentActiveChannel_(NULL)
{
  LOG_TRACE << "EventLoop created " << this << " in thread " << threadId_;
  if (t_loopInThisThread)
  {
    LOG_FATAL << "Another EventLoop " << t_loopInThisThread
              << " exists in this thread " << threadId_;
  }
  else
  {
    t_loopInThisThread = this;
  }
  **_wakeupChannel_->setReadCallback(
      boost::bind(&EventLoop::handleRead, this));**
  // we are always reading the wakeupfd
  wakeupChannel_->enableReading();
}

EventLoop::handleRead函数是没有参数的，而ReadEventCallback是带Timestamp参数的，这二者不匹配。

**Comment by chenshuo** [comment]
This is not a bug.

It's a feature of `boost::bind` and `std::bind`, as stated in its document [1]:

>  Any extra arguments are silently ignored, 

[1] https://www.boost.org/doc/libs/1_73_0/libs/bind/doc/html/bind.html

**Comment by icooperlee** [comment]
Thank you very much.  But i still don't understand, and this is a test.
------------------------------------------------------------------------
mybind.c

```
#include <stdio.h>
#include <boost/bind.hpp>
#include <boost/function.hpp>

typedef boost::function<void(int)> CallbackFun;

void myprint()
{
    printf("myprint...\n");
}

int main()
{
    CallbackFun f;
    f = boost::bind(myprint);
    f();

    return 0;
}

```
-----------------------------

$ g++ mybind.c -o bind
mybind.c: In function ‘int main()’:
mybind.c:17:7: error: no match for call to ‘(CallbackFun {aka boost::function<void(int)>}) ()’
     f();
       ^
In file included from /usr/include/boost/function/detail/maybe_include.hpp:22:0,
                 from /usr/include/boost/function/detail/function_iterate.hpp:14,
                 from /usr/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:52,
                 from /usr/include/boost/function.hpp:70,
                 from mybind.c:3:
/usr/include/boost/function/function_template.hpp:1048:7: note: candidate is:
 class function<BOOST_FUNCTION_PARTIAL_SPEC>
       ^
/usr/include/boost/function/function_template.hpp:758:17: note: boost::function1<R, T1>::result_type boost::function1<R, T1>::operator()(T0) const [with R = void; T0 = int; boost::function1<R, T1>::result_type = void]
     result_type operator()(BOOST_FUNCTION_PARMS) const

**Comment by icooperlee** [comment]
Sorry, i have a mistake. Should be like this.
```
#include <stdio.h>
#include <boost/bind.hpp>
#include <boost/function.hpp>


typedef boost::function<void(int)> CallbackFun;

void myprint()
{
    printf("myprint...\n");
}

int main()
{
    CallbackFun f;
    f = boost::bind(myprint);
    f(1);

    return 0;
}
```

**Comment by chenshuo** [comment]
What's the question here? The code works just fine: https://ideone.com/Mhjl02

```c++
#include <stdio.h>
#include <functional>
 
using Callback = std::function<void(int)>;
 
void print() {
	printf("Hello\n");
}
 
int main() {
	Callback f = std::bind(print);
	f(43);  // prints "Hello"
	return 0;
}
```

**Comment by icooperlee** [comment]
Now, I understand that is not a bug.   Can you tell me why it works  well? Thanks.   **"f = boost::bind(myprint); // ok.   why ?"** 
```
#include <stdio.h>
#include <boost/bind.hpp>
#include <boost/function.hpp>


typedef boost::function<void(int)> CallbackFun;

typedef void (*pFun)(int);


void myprint()
{
    printf("myprint...\n");
}


int main()
{
    CallbackFun f;
    f = boost::bind(myprint); // ok.   why ?
    f(1); // ok

    pFun pf = boost::bind(myprint); // complie error 

    return 0;
}
```

**Reply by chenshuo** [reply]
https://github.com/chenshuo/muduo/discussions/581#discussioncomment-2769084

**Comment by chenshuo** [comment]
https://github.com/chenshuo/muduo/discussions/581#discussioncomment-2769084
